mixnet/client/src/receiver.rs

```diff
@@ -15,40 +15,62 @@ pub struct Receiver;
 impl Receiver {
     pub async fn run(
         node_address: SocketAddr,
-        message_tx: impl Sink<Vec<u8>> + Unpin + Clone,
-    ) -> Result<(), Box<dyn Error>> {
-        let mut socket = TcpStream::connect(node_address).await?;
-        let mut message_reconstructor: MessageReconstructor = Default::default();
+    ) -> Result<impl Stream<Item = Result<Vec<u8>, Box<dyn Error>>> + Send + 'static, Box<dyn Error>>
+    {
+        let socket = TcpStream::connect(node_address).await?;
 
-        loop {
-            let body = Body::read(&mut socket).await?;
-            match body {
-                Body::SphinxPacket(_) => return Err("received sphinx packet not expected".into()),
-                Body::FinalPayload(payload) => {
-                    Self::handle_payload(payload, message_tx.clone(), &mut message_reconstructor)
-                        .await?
+        // MessageReconstructor buffers all received fragments (payloads)
+        // and eventually returns reconstructed messages.
+        let message_reconstructor: MessageReconstructor = Default::default();
+
+        Ok(stream::unfold(
+            (socket, message_reconstructor),
+            |(mut socket, mut message_reconstructor)| async move {
+                loop {
```

Comment:
[7e3c6d1](https://github.com/logos-co/nomos-node/pull/320/commits/7e3c6d1d315505a852af8ab714ca342a704cb137) I just added an infinite loop here because we may need to read multiple bodys from the socket until a message is reconstructed completely. For the same reason, we should declare the `message_reconstructor` outside of this closure and deal with it as the state for `stream::unfold` along with the `socket`.

Reply:
Wait why? We read the socket asynchronously which means we wait till we read a full body. What am I missing?

Reply:
> We read the socket asynchronously which means we wait till we read a full body.

True. The `Body::read(socket).await` returns a full body (payload). But, as a mixnet client, we need to gather multiple payloads until we can reconstruct the original message completely (because the original message was splitted into several payloads (fragments) by the message sender).

Reply:
Because we defined the stream to return reconstructed messages, our unfold closure shouldn't return anything until it gather all necessary payloads for resconstructing a message.

Reply:
In other words, we shouldn't consider [this](https://github.com/logos-co/nomos-node/pull/320/commits/7e3c6d1d315505a852af8ab714ca342a704cb137#diff-1382d0d8a9a6accf6066a9975609e83a69c2695ac23ce95c15ea58413cdd230eL52) as a error case.

Reply:
But that is a different approach, we can consume this stream to get the full message, doesn't mean we do it here.

Reply:
Yes, reconstruction is a mixclient responsibility. What I meant is that we can take this stream and do the reconstruction **consuming** this stream.  So we go from a `Stream<Vec<u8>>` to `Stream<Message>`, separating responsibilities as well.

Reply:
I think I don't get it. Does it mean that we make this stream as `Stream<Payload>` and make the other `Stream<Message>` which consumes the first stream?

Reply:
Thank you. It makes sense. It sounds more flexible than putting everything in a single stream with the infinite loop. I’ll try it.

Reply:
[03b14f6](https://github.com/logos-co/nomos-node/pull/320/commits/03b14f624c6a10e9918f7dde02733c075b491578) I just splitted the role into two separated streams for better flexibility, but I would like to ask if this is what you meant, because I feel the overall internal behaviour is not very different from my previous implementation.

Reply:
Looks good to me. We do not know how many fragments a message is constructed by right? 
Also, is not clear to me what would happend if you receive diferent chunks for diferent messages in the stream (is that even possible?).